test/one/profiler/test/TestProcess.java

```diff
@@ -0,0 +1,367 @@
+            
+            File profileDirectory = tmpFiles.get("%f");
+            if (profileDirectory != null) {
+                moveLog(profileDirectory, directory, "profile");
+            }
+
+        } catch (Throwable e) {
+            log.log(Level.WARNING, "Failed to retain logs: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void close() {
+        p.destroy();
+        
+        try {
+            waitForExit(p, 20);
+        } catch (TimeoutException e) {
+            log.log(Level.WARNING, "Failed to terminate child process", e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } finally {
+            if (testName != null) {
+                moveLogs();
+            }
+            clearTempFiles();
+        }
+    }
+
+    public long pid() {
+        try {
+            return (long) pid.invokeExact(p);
+        } catch (Throwable e) {
+            throw new IllegalStateException("Unsupported API", e);
+        }
+    }
+
+    public Output waitForExit(String fileId) throws TimeoutException, InterruptedException, IOException {
+        waitForExit(p, timeout);
+        return readFile(fileId);
+    }
+
+    private void waitForExit(Process p, int seconds) throws TimeoutException, InterruptedException {
+        if (!p.waitFor(seconds, TimeUnit.SECONDS)) {
+            p.destroyForcibly();
+            throw new TimeoutException("Child process has not exited");
+        }
+    }
+
+    public Output profile(String args) throws IOException, TimeoutException, InterruptedException {
+        return profile(args, false);
+    }
+
+    public Output profile(String args, boolean sudo) throws IOException, TimeoutException, InterruptedException {
+        List<String> cmd = new ArrayList<>();
+        if (sudo) {
+            cmd.add("/usr/bin/sudo");
+        }
+        cmd.add("build/bin/asprof");
+        addArgs(cmd, args);
+        cmd.add(Long.toString(pid()));
+        log.log(Level.FINE, "Profiling " + cmd);
+        
+        String extension = getExtFromArgs(args);
+        if (args.contains("%f")) {
+            extension = "";
+        }
+        Process p = new ProcessBuilder(cmd)
+                    .redirectOutput(createTempFile("%pout", extension))
+                    .redirectError(createTempFile("%perr", null))
+                    .start();
+
+        waitForExit(p,10);
+        int exitCode = p.waitFor();
+        if (exitCode != 0) {
+            throw new IOException("Profiling call failed " + readPOut().toString() + readPErr().toString());
+        }
+
+        return readPOut();
+    }
+
+    public Output readFile(String fileId) throws IOException {
+        File f = tmpFiles.get(fileId);
+        String fileExtension = f.getName().replaceAll(".*\\.(\\w+)$", "$1");
+        if (fileExtension.equals("jfr")) {
+            return readJfr(f);
+        } else {
+            return readFile(f);
+        }
```

Comment:
That's a questionable solution. A test usually knows whether it expects JFR or Text output. Why not calling `readFile` vs. `readJfrFile` explicitly? They can return different objects: `Output` vs. `JfrOutput`, thus resolving my previous concern about embedding `JfrOutput` in `Output`.